### DISCUSSION TITLE
Get trial/iteration number inside a timeline

### DISCUSSION THREAD

**Discussion by mrinmayik** [jspsych/jsPsych#2653] [Q&A] [answered]
Hello jsPsych experts,

I have a question that I think has a simple solution, but I can't seem to figure it out! I have an experiment where stimuli are displayed in a timeline using a timeline variable. In each trial I add certain information to the output data to be used during data analysis. Here, I want to add information about trial number, i.e., which iteration of the timeline variable is being currently implemented. How do I do this?

Here's my timeline:
```
var EncodingProcedure = {
    timeline: [
        {
            type: jsPsychImageKeyboardResponse,
            stimulus: jsPsych.timelineVariable('stimuli'),
            choices: ['i', 'o'],
            trial_duration: 2000,
            response_ends_trial: false,
            data: {
                counterbalance: CBNum,
                task: 'encoding',
                run: run+1,
                percent: jsPsych.timelineVariable('percent'),
                squaredim: jsPsych.timelineVariable('squaredim'),
                displacement: jsPsych.timelineVariable('displacement'),
                listnum: jsPsych.timelineVariable('listnum'),
                trialtype: jsPsych.timelineVariable('trialtype'),
                trialnum: //HERE I WANT TO ADD THE TRIAL NUMBER
                imagename: jsPsych.timelineVariable('imagename'),
            },
            css_classes: ['alignverticalhorizontal']
        }
    ],
    timeline_variables: StimList[BName].Encoding
}
```

Quick note: I know that a column named trial_index is automatically added. But it seems to number every single event in the experimental timeline (including preloading, etc.). I want to only number the event where the stimulus is displayed. 

In the worst case I can change `StimList[BName].Encoding` to have a trialnum variable and add it to the data using `jsPsych.timelineVariable`. But I was wondering if there's an easier way!

Thank you,
Mrinmayi

**Comment by ChristopheBossens** [comment, marked as answer]
The following approach seems to work:
1. Create a separate `trial_counter `variable. 
2. Add an `on_finish `function in which you increment the value of the `trial_counter `variable and add it to the trial data. I noticed that you need to use the `on_finish `function. Simply adding it to the data key does not show an incremented value.


```
let trial_counter  = 0;
  let face_name_procedure = {
      timeline: [
          {
              type: jsPsychHtmlKeyboardResponse,
              stimulus: '+',
              choices: "NO_KEYS",
              trial_duration: 500
          },
          {
              type: jsPsychImageKeyboardResponse,
              stimulus: jsPsych.timelineVariable('face'),
              choices: "NO_KEYS",
              trial_duration: 2500,
              on_finish: function(data){
                  trial_counter += 1;
                  data.trial_counter = trial_counter;
              },

          }
      ],
      timeline_variables: [
          { face: 'person-1.jpg' },
          { face: 'person-2.jpg' },
          { face: 'person-3.jpg' },
          { face: 'person-4.jpg' }
      ]
  }
```

**Reply by mrinmayik** [reply]
Yes that worked!! Thanks so much! :)
